### PR TITLE
Update changelog in preparation for 1.2.0 release

### PR DESCRIPTION
**Pull request by Technius**
Had lots of time on a long international flight, so I decided to write up the changelog. Looking back, there's a lot of changes that have been added over the past year and a half since 1.1.0's release.

@xournalpp/core I'm sure there are obvious mistakes and omissions in this list, so please take a look and point out any problems.

Last checked against: `d3cd981aa94843ea6d32f532659970eaa1568988`

### CODE REVIEW

**Review by rolandlo — COMMENTED** [review]
Thanks @Technius for taking the time to write the Changelog for 1.2.0. There are really a lot of changes/improvements.

Here are some comments after a first read through. There seem to be only few typos.

Do we really drop 18.04 *before* the 1.2.0 release? Why? I thought we will do so only after the 1.2.0 release.

We should also mention that on MacOS the Gtk version was updated to 3.24.35 which solves the lag issue.

**Comment by Technius** [comment]
> Do we really drop 18.04 before the 1.2.0 release? Why? I thought we will do so only after the 1.2.0 release.

It's not 100% necessary, but:
* My main concern is that older Ubuntus and Debians do not fully support C++17 filesystem, so they cannot be built in PPAs or the official repository (as they need some sort of downloaded C++17 shim).
* Because Ubuntu 18.04 will be removed from GHA (see #4324), we'll need to modify the CI scripts to build in Docker for Ubuntu 18.04 if we want to continue support. Not a big deal, but we could avoid all this work by dropping support altogether.
* Ubuntu 18.04 will be EOL in April 2023.

These are not major blockers, so I can remove that line if you disagree with me.

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,172 @@
...
+  * Minor performance improvements (#4004).
+* Floating toolbox
+  * Added capability to open the floating toolbox with a secondary button
+    (#3270).
+* Packaging
+  * Added build flags to force or disable X11 linking (#4146).
+  * AppImage: added support for AppImageUpdate (#1915, #4265).
+* Miscellaneous
+  * Improved file size of documents with large amonuts of stroke data by up to
+    15% (#3831, #4065).
```
Comment (rolandlo) [inline CHANGELOG.md:126]: That looks like the wrong issue. I didn't find out which one was meant. Also there is a typo.
```suggestion
  * Improved file size of documents with large amounts of stroke data by up to
    15% (#4065).
```

**Review by rolandlo — COMMENTED** [review]
Some more comments.

**Comment by rolandlo** [comment]
> > Do we really drop 18.04 before the 1.2.0 release? Why? I thought we will do so only after the 1.2.0 release.
> 
> It's not 100% necessary, but:
> 
>     * My main concern is that older Ubuntus and Debians do not fully support C++17 filesystem, so they cannot be built in PPAs or the official repository (as they need some sort of downloaded C++17 shim).
> 
>     * Because Ubuntu 18.04 will be removed from GHA (see [Update CI build environments #4324](https://github.com/xournalpp/xournalpp/issues/4324)), we'll need to modify the CI scripts to build in Docker for Ubuntu 18.04 if we want to continue support. Not a big deal, but we could avoid all this work by dropping support altogether.
> 
>     * Ubuntu 18.04 will be EOL in April 2023.
> 
> 
> These are not major blockers, so I can remove that line if you disagree with me.

Well, I would just release 1.2.0 for Ubuntu 18.04 on the Github page, not worry about PPA's and drop support right afterwards. It doesn't create any efforts to us and it still helps users of Ubuntu 18.04 (and derived distros) who have waited for the stable release.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,172 @@
...
+* Stroke and eraser tools
+  * Added a system cursor option (#3540).
+  * Added stoke cap styles to the file format (#3326).
+  * Added a "double ended arrow" stroke style (#2362, #3946).
+  * Added a Preferences setting for setting the minimum size of the shape
+    recognizer (#4283).
+  * Added a toolbar item for changing fill opacity (#4263).
+  * Improved minor graphical aspects of the fill opacity dialog (#3515).
+  * Improved performance by rendering strokes with integer coordinates (#4080).
+  * Rewrote the stroke rendering code to fix bugs and improve performance
```
[the remark this answers is not in the document]
Comment (Technius) [reply CHANGELOG.md:59]: #4304 makes some optimization changes, so I think it is correct. #3580 fixes a regression, but it's appropriate to add nonetheless.

File: CHANGELOG.md (hunk 1 above)
Comment (Technius) [reply CHANGELOG.md:126]: You're right, #3831 should go on the next line.

**Comment by Technius** [comment]
The line mentioning "Ubuntu 18.04 support dropped" has been removed.

**Comment by Technius** [comment]
This needs to be updated with the past four months of changes.

**Comment by Technius** [comment]
Updated again to track changes up to `0d132dfd90d5d039fd923b6289314b306b2ad16e`.

* Not sure if #4505 is a regression or existing bug.
* #2897 still needs a UI implementation.
* We will want to fix #4811

**Comment by rolandlo** [comment]
I have updated the changelog with the changes since April 24. It also includes #4962 which hasn't been merged yet.

**Comment by rolandlo** [comment]
I went through the whole changelog and checked every issue number. With the latest minor additions and corrections everything looks fine to me.

Squashing and merging in 24 hours if no objections are raised.

**Comment by NichopeJustPassing** [comment]
> Well, I would just release 1.2.0 for Ubuntu 18.04 on the Github page, not worry about PPA's and drop support right afterwards. It doesn't create any efforts to us and it still helps users of Ubuntu 18.04 (and derived distros) who have waited for the stable release.

Hi! I am one of such users, just checking: In the release for 1.2.0, there is no "*-Ubuntu-bionic*" files. Did the situation change, regarding 1.2.0 being the final release for 18.04? Or am I missing something?

Btw: Thanks a lot for Xournal++ (to both of you). It has become one of my favourite programs over the last year (I wish I had known about it during my bachelor's), but I did not remember that I had a GitHub profile, so I never got to comment here 😅

**Comment by rolandlo** [comment]
> > Well, I would just release 1.2.0 for Ubuntu 18.04 on the Github page, not worry about PPA's and drop support right afterwards. It doesn't create any efforts to us and it still helps users of Ubuntu 18.04 (and derived distros) who have waited for the stable release.
> 
> Hi! I am one of such users, just checking: In the release for 1.2.0, there is no "_-Ubuntu-bionic_" files. Did the situation change, regarding 1.2.0 being the final release for 18.04? Or am I missing something?
> 
It took longer to release version 1.2.0 than we expected. The Ubuntu 18.04 image was then already dropped by Azure Pipelines, where we build the releases, and so we had to drop support for Ubuntu 18.04 already for version 1.2.0.
If you're still using Ubuntu 18.04 (or a derivative), consider using the [flatpak](https://flathub.org/apps/com.github.xournalpp.xournalpp) or [building from source](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md).

> Btw: Thanks a lot for Xournal++ (to both of you). It has become one of my favourite programs over the last year (I wish I had known about it during my bachelor's), but I did not remember that I had a GitHub profile, so I never got to comment here 😅

Glad you like it!

**Comment by NichopeJustPassing** [comment]
Thanks for the quick reply!

> It took longer to release version 1.2.0 than we expected. The Ubuntu 18.04 image was then already dropped by Azure Pipelines, where we build the releases, and so we had to drop support for Ubuntu 18.04 already for version 1.2.0. If you're still using Ubuntu 18.04 (or a derivative), consider using the [flatpak](https://flathub.org/apps/com.github.xournalpp.xournalpp) or [building from source](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md).

Oh, that's unfortunate, but on the other hand "migrating" away from the .deb release also means that I'll be able to get all bugfixes in the following releases, and whichever new features come later (really looking forward to shortcut customisation or at least spacebar-activated autohand).
I imagine that given the dropping of support and likely divergence in dependency package versions, installing from flatpak is less likely to cause problems later on, right?

**Comment by rolandlo** [comment]
Yes, installing from flatpak is the more convenient option.

What do you mean by "(spacebar-activated) autohand"?

**Comment by NichopeJustPassing** [comment]
> Yes, installing from flatpak is the more convenient option.
Great! Thanks for the confirmation

> What do you mean by "(spacebar-activated) autohand"?
I meant the function to "grab" the canvas by pressing the spacebar (like the middle button on the mouse does), to allow for flexible navigation on pen tablets while keeping the stylus in the air, like in many drawing programs.
I think I saw a feature request about it last year, but if I remember correctly it was postponed until deeper changes were made to the codebase to allow for more flexible shortcuts, or something like that.

On Windows, I circumvent the issue by using the Wacom properties to make one of the stylus buttons be a middle click, and make Xournal++ treat one of the stylus inputs as a mouse (and the rest as a pen), and it works with pressure sensitivity, at the cost of having a mouse pointer for the pen. But on Linux I have not found a way to do it yet (although that may be a limitation of my system, which only lets me treat the buttons as "stylus buttons")

**Pull request merged**